### PR TITLE
Remove deprecations to support ember 4.x

### PR DESCRIPTION
**Pull request by fsmanuel**
Ref https://github.com/ember-intl/cp-validations/issues/85

### CODE REVIEW

File: addon/validators/messages.js [hunk 1]
```diff
@@ -43,27 +42,35 @@ export default ValidatorsMessages.extend({
...
 
     if (intl && intl.exists(key)) {
       return this.formatMessage(intl.t(key, options));
     }
 
-    this._warn(`[@ember-intl/cp-validations] Missing translation for validation key: ${key}\nhttp://offirgolan.github.io/ember-cp-validations/docs/messages/index.html`, false, {
-      id: 'ember-intl-cp-validations-missing-translation'
-    });
+    this._warn(
+      `[@ember-intl/cp-validations] Missing translation for validation key: ${key}\nhttp://offirgolan.github.io/ember-cp-validations/docs/messages/index.html`,
```
Comment (tehhowch) [inline addon/validators/messages.js]: URL should be to adopted addons now, i believe

File: config/ember-try.js [hunk 2]
```diff
@@ -1,83 +1,93 @@
...
+            'ember-source': await getChannelURL('beta'),
+          },
+        },
+      },
+      {
+        name: 'ember-canary',
+        npm: {
+          devDependencies: {
+            'ember-source': await getChannelURL('canary'),
+          },
```
Comment (tehhowch) [inline config/ember-try.js]: These (and the embroider scenarios) should declare the needed dev deps in the scenario config:
```js
npm: {
  devDependencies: {
    'ember-source': ....
    'ember-auto-import': '^2',
    'webpack': '^5',
  },
},
```

File: .template-lintrc.js [hunk 3]
```diff
@@ -0,0 +1,9 @@
+'use strict';
+
+module.exports = {
+  extends: 'recommended',
+  rules: {
+    'no-curly-component-invocation': false,
+    'no-implicit-this': false,
```
Comment (tehhowch) [inline .template-lintrc.js]:
```suggestion
    'no-curly-component-invocation': 'warn',
    'no-implicit-this': 'warn',
```

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -1,76 +1,79 @@
...
+          - ember-4.0
           - ember-release
           - ember-beta
-        experimental: [false]
-        include:
-          - scenario: ember-canary
-            experimental: true
+          - ember-canary
+          # - embroider-safe
+          # - embroider-optimized
```
Comment (tehhowch) [inline .github/workflows/ci.yml]: you can enable these with the changes i mention in the ember try file

File: .template-lintrc.js (hunk 3 above)
Comment (fsmanuel) [reply .template-lintrc.js]: Good catch! I updated the template and removed the rules.

File: .github/workflows/ci.yml (hunk 4 above)
Comment (fsmanuel) [reply .github/workflows/ci.yml]: embroider-optimized fails: https://github.com/ember-intl/cp-validations/actions/runs/3096110587/jobs/5011285918

**Comment by fsmanuel** [comment]
@ijlee2 @tehhowch everything except `embroider-optimized` is now green.

File: tests/dummy/app/templates/application.hbs [hunk 5]
```diff
@@ -2,14 +2,47 @@
...
-<div class="username-validation">{{model.validations.attrs.username.message}}</div>
-
-{{input id="email" value=model.email}}
-{{input id="emailConfirmation" value=model.emailConfirmation}}
-{{input id="password" value=model.password}}
-{{input id="passwordConfirmation" value=model.passwordConfirmation}}
-{{input id="username" value=model.username}}
+<input
+  id="email"
+  value={{this.model.email}}
```
Comment (ijlee2) [inline tests/dummy/app/templates/application.hbs]: **suggestion (optional)**

For accessibility, I'd recommend specifying the `type`. For example, `type="email"`, `type="password"`, and `type="text"`. Specifying the type can also help us avoid re-implementing basic validations.

We can also consider adding a label to each input and using the `for` attribute to tie the label to the input.

File: tests/dummy/app/controllers/application.js [hunk 6]
```diff
@@ -0,0 +1,10 @@
+import Controller from '@ember/controller';
+import { action } from '@ember/object';
+
+export default class extends Controller {
+  @action
+  update(attribute, event) {
+    let value = event.target.value;
+    return this.model.set(attribute, value);
```
Comment (ijlee2) [inline tests/dummy/app/controllers/application.js]: It feels a bit iffy to modify `model` (I like to think of what the `model` hook returns as read-only), but I understand that the PR is focused on updating the current syntax.

I might suggest that, in the future, we look at defining the validations and and the form "object" directly in the controller (currently these are defined in `tests/dummy/app/models/dummy.js`).

**Review by ijlee2 — APPROVED** [review]
✨

File: tests/dummy/app/models/dummy.js [hunk 7]
```diff
@@ -1,40 +1,38 @@
...
 });
 
-export default Ember.Object.extend(Validations, {
+// eslint-disable-next-line ember/no-classic-classes
+export default EmberObject.extend(Validations, {
   username: '',
   password: '',
-  email: ''
+  email: '',
 });
```
Comment (tehhowch) [inline tests/dummy/app/models/dummy.js]: You can use native syntax here:

```js
export default class extends EmberObject.extend(Validations) {
  username = '';
  ...
}
```

File: .github/workflows/ci.yml (hunk 4 above)
Comment (tehhowch) [reply .github/workflows/ci.yml]: Even when given the configuration?

```
embroiderOptimized({
  npm: {
    devDependencies: {
      'ember-auto-import': '^2',
      webpack: '^5',
    },
  },
}),
```
I wonder if it is from something upstream..

Comment (fsmanuel) [reply .github/workflows/ci.yml]:
```js
const specifier = '@ember-intl/cp-validations@validator:messages';
const instance = this.owner.lookup(specifier);
instance.getMessageFor('foobarbaz');
```

seems like `instance` is undefined in`embroider-optimized` scenario.

```js
function embroiderOptimized(extension) {
    return extendScenario({
        name: 'embroider-optimized',
        npm: {
            devDependencies: {
                '@embroider/core': currentEmbroiderVersion,
                '@embroider/webpack': currentEmbroiderVersion,
                '@embroider/compat': currentEmbroiderVersion,
                // Webpack is a peer dependency of `@embroider/webpack`
                webpack: '^5.0.0',
            },
        },
        env: {
            EMBROIDER_TEST_SETUP_OPTIONS: 'optimized',
        },
    }, extension);
}
```

I think `ember-auto-import` `v2` is present via `@embroider/util` which has `"ember-auto-import": "^2.2.0",` as a dependency. Also the `embroider-safe` scenario works which has the same `devDependencies` as `embroiderOptimized` from above.

```
$ npm ls @embroider/util
@ember-intl/cp-validations@5.0.0 /Users/manuel/Code/opensource/cp-validations
└─┬ @ember/test-helpers@2.8.1
  └── @embroider/util@1.8.3
```

I'll create a new issue with the information. If you want to try to fix it a PR is very welcome.

Comment (fsmanuel) [reply .github/workflows/ci.yml]: Tracked in https://github.com/ember-intl/cp-validations/issues/92

File: tests/dummy/app/templates/application.hbs (hunk 5 above)
Comment (fsmanuel) [reply tests/dummy/app/templates/application.hbs]: https://github.com/ember-intl/cp-validations/pull/90/commits/2b569ed725d96fb7b2720c6f3644660b0a3a3d85

File: tests/dummy/app/controllers/application.js (hunk 6 above)
Comment (fsmanuel) [reply tests/dummy/app/controllers/application.js]: Happy to do a review if you open a PR =)

File: tests/dummy/app/models/dummy.js (hunk 7 above)
Comment (fsmanuel) [reply tests/dummy/app/models/dummy.js]: https://github.com/ember-intl/cp-validations/pull/90/commits/2b569ed725d96fb7b2720c6f3644660b0a3a3d85

**Pull request merged**